### PR TITLE
Upgrade helper

### PR DESCRIPTION
**Pull request by sgallagher**
Add Modulemd.UpgradeHelper object
    
This is the implementation of an object that will hold known
module:stream values to provide a way to resolve ambiguous
upgrades from ModuleStreamV2 dependencies.

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

### CODE REVIEW

**Comment by sgallagher** [comment]
This is the basic object; I still need to work on associating it with the `ModuleIndex` and `Module` objects. But if you could review the object in the meantime, that would be helpful.

File: modulemd/include/modulemd-2.0/modulemd-upgrade-helper.h [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+
+/**
+ * modulemd_upgrade_helper_get_known_streams_as_strv: (rename-to modulemd_upgrade_helper_get_known_streams)
+ * @self: (in): This #ModulemdUpgradeHelper
+ * @module_name: (in): The name of the module to return a list of known streams
+ * for.
+ *
+ * Returns: (transfer full): A list of known streams to provide clues to the
+ * stream upgrade process.
```
Comment (mmathesius) [inline modulemd/include/modulemd-2.0/modulemd-upgrade-helper.h:96]:
```suggestion
 * stream upgrade process. NULL if there are no known streams for @module_name.
```
It's probably worth adding a quick test that querying an unknown module returns NULL, too.

**Review by mmathesius — COMMENTED** [review]
Just one nit-pick on documentation and a suggested test to add.

File: modulemd/include/modulemd-2.0/modulemd-upgrade-helper.h (hunk 1 above)
Comment (sgallagher) [reply modulemd/include/modulemd-2.0/modulemd-upgrade-helper.h:96]: I went the other way; it turns out that PyGObject always returns `[ ]` there, even if it gets NULL back. I decided to keep it consistent and I'll return a zero-length array to C.

**Comment by sgallagher** [comment]
I fixed up the requested issue and added the test.

I also included two other patches, another one for fixing up Travis tests (since this PR showed it was still not picking up the overrides properly) and also clean up some memory issues detected by clang.

File: modulemd/modulemd-yaml-util.c [hunk 2]
```diff
@@ -504,6 +504,17 @@ modulemd_yaml_parse_string_set (yaml_parser_t *parser, GError **error)
       yaml_event_delete (&event);
     }
 
+  /* Work around false-positive in clang static analysis which thinks it's
+	   * possible for this function to return NULL and not set error.
+	   */
```
Comment (mmathesius) [inline modulemd/modulemd-yaml-util.c:509]: Odd comment indentation...

**Review by mmathesius — APPROVED** [review]
Aside from the noted odd comment indentation, this looks good to me. Thanks!

File: modulemd/modulemd-yaml-util.c (hunk 2 above)
Comment (sgallagher) [reply modulemd/modulemd-yaml-util.c]: Fixed

File: modulemd/modulemd-yaml-util.c [hunk 3]
```diff
@@ -504,6 +504,17 @@ modulemd_yaml_parse_string_set (yaml_parser_t *parser, GError **error)
       yaml_event_delete (&event);
     }
 
+  /* Work around false-positive in clang static analysis which thinks it's
+	 * possible for this function to return NULL and not set error.
+	 */
```
Comment (mmathesius) [inline modulemd/modulemd-yaml-util.c:509]: Indentation is still weird--caused by a hard tab at the beginning of those two lines.

Comment (sgallagher) [reply modulemd/modulemd-yaml-util.c]: Thanks, it looked fine on my editor because the tabs were four spaces wide. Fixed now.

Comment (mmathesius) [reply modulemd/modulemd-yaml-util.c]: No problem. I thought `clang-format` expanded hard tabs...

Comment (sgallagher) [reply modulemd/modulemd-yaml-util.c]: It doesn't touch tabs that are inside a comment block.
